Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81763

These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @ezyang @ngimel @peterbell10

Differential Revision: [D42354776](https://our.internmc.facebook.com/intern/diff/D42354776)

lezcano committed:
Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them

These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

Lezcano commented:
About BC compatibility. We run the same test as described in https://github.com/pytorch/pytorch/pull/81761#issuecomment-1190249259

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2682,94 +2682,92 @@ Tensor& linalg_norm_out(const Tensor& X, c10::string_view ord, OptionalIntArrayR
-      result = at::sqrt(at::sum((self * self), dim_, keepdim));
-    }
+  auto device = self.device();
+  if (self.layout() == Layout::Strided && (device == kCPU || device == kCUDA || device == kMeta)) {
```
nit - there are 6 (?) copies of this check with minor variations - time for a utility function?

Lezcano replied:
Fair enough

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge -f "unrelated flaky tests"

mehtanirav commented:
This is causing some internal tests to fail with the following error.

```
terminate called after throwing an instance of 'c10::Error'
  what():  Could not find schema for aten::linalg_vector_norm.out
  
  
Exception raised from findSchemaOrThrow at xplat/caffe2/aten/src/ATen/core/dispatch/Dispatcher.cpp:131 (most recent call first):
(no backtrace available)
```

mehtanirav commented:
@pytorchbot revert -m "Internal breakages" -c ghfirst

kit1980 commented:
I'll help to make sure next the merge will not break Meta-internal stuff.

lezcano commented:
Thank you so much @kit1980

lezcano commented:
Anything you need from my end just ping me

kit1980 commented:
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42354776).

lezcano commented:
What's the state of this @kit1980?

kit1980 commented:
@lezcano Meta-internal mobile tests are failing. I'm fixing.

lezcano commented:
Thank you!

kit1980 commented:
@pytorchbot rebase

kit1980 commented:
@lezcano Could you rebase?

lezcano committed:
Update on "Dispatch the auxiliary frobenius_norm and nuclear_norm to better implementations and deprecate them"


These functions will be legacy functions. We deprecate them, but we also
take this chance to dispatch to a more efficient and consistent implementation.
Doing so should help writing a conversion rule for these to be able to
remove them once and for all

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano ezyang ngimel peterbell10

Differential Revision: [D42354776](https://our.internmc.facebook.com/intern/diff/D42354776)

[ghstack-poisoned]

lezcano commented:
@kit1980 done!

kit1980 commented:
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42354776).

huydhn commented:
@pytorchbot revert -c 'This breaks XLA test https://hud.pytorch.org/pytorch/pytorch/commit/34e8eb229db76f7f5eb8f18c062dbd1ee47f8b12' -c landrace

huydhn commented:
@pytorchbot revert -m 'This breaks XLA test https://hud.pytorch.org/pytorch/pytorch/commit/34e8eb229db76f7f5eb8f18c062dbd1ee47f8b12' -c landrace

kit1980 commented:
Here is the XLA test that needs to be updated: https://github.com/pytorch/xla/blob/master/test/cpp/test_aten_xla_tensor.cpp#L1887

kit1980 commented:
Or we can just delete the test as did https://github.com/pytorch/xla/pull/3976
@lezcano what do you think?

lezcano commented:
uh, no, we should fix that. Odd that that happened, ci was green...

lezcano commented:
Right, this is an issue with some new symint stuff. I'll look into that.

kit1980 commented:
@lezcano don't reopen this PR, we didn't actually revert this.
Create a new PR to forward-fix the XLA test failure.

lezcano commented:
Sure thing, will do that on Monday.
